terraform_ci/terraform_cd.py

```diff
@@ -68,49 +113,55 @@ def terraform_cd(**kwargs):
+        if kwargs["include_artifacts"]:
+            LOG.debug("Storing archive under %s", release_archive_full_path)
+            symlink(getcwd(), osp.join(tmp_dir, module_directory_name))
 
-        with open(osp.join(tmp_dir, release_archive), "wb") as archive_descriptor:
             proc = Popen(
                 [
-                    "git",
-                    "archive",
-                    "--format=tar.gz",
-                    "--prefix={project}-{tag}/".format(project=module_name, tag=tag),
-                    tag,
-                ],
-                stdout=archive_descriptor,
+                    "tar",
+                    "--directory={tmp}".format(tmp=tmp_dir),
+                    "-chzf",
+                    release_archive_full_path,
+                    module_directory_name,
+                ]
             )
-        proc.communicate()
+            LOG.debug("Running %s", proc.args)
+            proc.communicate()
+
+        # generate archive using git archive
+        else:
+            with open(release_archive_full_path, "wb") as archive_descriptor:
+                proc = Popen(
+                    [
+                        "git",
+                        "archive",
+                        "--format=tar.gz",
+                        "--prefix={project}-{tag}/".format(
+                            project=module_name, tag=tag
+                        ),
+                        tag,
+                    ],
+                    stdout=archive_descriptor,
+                )
+            LOG.debug("Running %s", proc.args)
+            proc.communicate()
 
+        # sync tar.gz to s3
         setup_environment(
             config_path=kwargs["env_file"], role=kwargs["aws_assume_role_arn"]
         )
 
-        try:
-            s3_client = boto3.client("s3")
-
-            with open(osp.join(tmp_dir, release_archive), "rb") as archive_descriptor:
-                s3_client.upload_fileobj(
-                    archive_descriptor,
-                    bucket,
-                    osp.join(module_name, release_archive),
-                    ExtraArgs={"ACL": "bucket-owner-full-control"},
-                )
-                LOG.info(
-                    "Published artifact to s3://%s/%s",
-                    bucket,
-                    osp.join(module_name, release_archive),
-                )
-
-        except ClientError as err:
-            LOG.error(err)
-            try:
-                sts_client = boto3.client("sts")
-                LOG.error("AWS caller: %s", sts_client.get_caller_identity()["Arn"])
+        send_to_s3(
+            bucket=bucket,
+            local_file=release_archive_full_path,
+            target_file=osp.join(module_name, release_archive),
+        )
 
-            except ClientError:
-                LOG.warning(
-                    "Failed to get AWS caller. Probably the client is not authenticated."
-                )
 
-            sys.exit(1)
+if __name__ == "__main__":
```

Comment:
the supported interface is only console entry points. let's not provoke users with other ways to call this.